integration/experiment/system_characterization/gen_cpu_activity_constconfig_recommendation.py

```diff
@@ -0,0 +1,259 @@
+
+    return df_filtered
+
+def system_memory_bandwidth_characterization(df_region_group, region, freq_col):
+    """
+    Perform characterization of the system memory bandwidth.  This
+    is used by the agent to help determine the appropriate uncore frequency
+    """
+    df = df_region_group.get_group(region)
+    uncore_freq_set = sorted(set(df[freq_col].to_list()))
+
+    mem_bw_dict = {}
+    for k in uncore_freq_set:
+        mem_df = df.groupby(freq_col).get_group(k)
+        mem_bw_dict[k] = mem_df['MSR::QM_CTR_SCALED_RATE'].mean()
+
+    return mem_bw_dict
+
+def frequency_recommendation(df_region_group, region, freq_col, energy_margin):
+    energy_col = 'package-energy (J)'
+
+    # Start with a region analysis for energy efficiency
+    df = df_region_group.get_group(region)
+    domain_freq_efficient = util.energy_efficient_frequency(df, freq_col, energy_col, energy_margin)
+
+    return domain_freq_efficient
+
+def get_config_from_frequency_sweep(full_df, region_list, mach,
+                                    core_energy_margin, uncore_energy_margin,
+                                    use_freq_req, hostname):
+    """
+    The main function. full_df is a report collection dataframe, region_list
+    is a list of regions to include.
+    """
+    df = extract_columns(full_df, region_list)
+
+    if df['frequency (Hz)'].isna().any() or df['uncore-frequency (Hz)'].isna().any():
+        raise RuntimeError('NaNs detected in the input report files')
+
+    df_region_group = df.groupby('region')
+
+    #Round entries to nearest step size
+    frequency_step = mach.frequency_step()
+    df['frequency (Hz)'] = (df['frequency (Hz)'] / frequency_step).round(decimals=0) * frequency_step
+    df['uncore-frequency (Hz)'] = (df['uncore-frequency (Hz)'] / frequency_step).round(decimals=0) * frequency_step
+
+    if use_freq_req:
+        freq_col = 'requested uncore-frequency (Hz)'
+    else:
+        freq_col = 'uncore-frequency (Hz)'
+
+    # Characterize MBM metrics using the uncore sensitive
+    # region (intensity_1)
+    mem_bw_characterization = system_memory_bandwidth_characterization(df_region_group,
+                                                                       region=region_list[0], freq_col=freq_col)
+
+    # A multi-step approach is used.
+    # First analyze the uncore sensitive
+    # region (ex: intensity_1) to find the efficient
+    # uncore frequency
+    uncore_freq_recommendation = frequency_recommendation(df_region_group, region=region_list[0],
+                                                          freq_col=freq_col, energy_margin=uncore_energy_margin)
+
+    # Round to step size
+    uncore_freq_recommendation = round(uncore_freq_recommendation / frequency_step) * frequency_step
+
+    # Then analyze the core senstivite region (ex: intensity_16)
+    # to find the most efficient core frequency
+    # when running at the uncore_freq_efficient determined
+    # above
```

Comment:
Is it more effective to choose maximum uncore frequency when determining efficient core frequency for a computationally bound benchmark? That way you could avoid any potential that the uncore frequency is low enough to lead to a bottleneck, which may cause the computational efficient frequency to appear lower than it truly is. This is likely an algorithmic nitpick, but I'm curious about the logic here.